scirpy/_tools/_clonotype_imbalance.py

```diff
@@ -0,0 +1,294 @@
+            )
+
+    # Convert records to data frames
+    clt_freq = pd.DataFrame.from_records(
+        clt_freq,
+        columns=[
+            target_col,
+            additional_hue,
+            groupby,
+            replicate_col,
+            "Normalized abundance",
+        ],
+    )
+    clt_stats = pd.DataFrame.from_records(
+        clt_stats, columns=[target_col, "pValue", "logpValue", "logFC"]
+    )
+    clt_stats = clt_stats.sort_values(by="pValue")
+
+    if inplace:
+
+        # Store calculated data
+        adata.uns[added_key] = {"abundance": clt_freq, "pvalues": clt_stats}
+        return
+
+    else:
+        return clt_freq, clt_stats
+
+
+def _create_case_control_groups(
+    df: pd.DataFrame,
+    replicate_col: str,
+    groupby: str,
+    additional_hue: Union[None, str, bool],
+    case_label: str,
+    control_label: Union[None, str],
+) -> List:
+    """Creates groups for comparison. 
+    
+    Parameters
+    ----------
+    df
+        A pandas dataframe with all the columns we will use for grouping.       
+    replicate_col
+        Column with batch or sample labels.
+    groupby
+        The column containing categories that we want to compare and find imbalance between
+    additional_hue
+        An additional grouping factor. If the `case_label` was tumor for example, this could
+        help make a distinction between imbalance in lung and colorectal tumors.
+    case_label
+        The label in `groupby` column that we want to compare. 
+    control_label
+        The label in `groupby` column that we use as a baseline for comparison. If not set
+        (None by default), all labels that are not equal to `case_label` make up the baseline. 
+    target_col
+        The clusters (clonotypes by default) that are imbalanced. 
+
+    Returns
+    -------
+    A list, where each item consist of a hue, list of cases, list of controls,
+    number of cases, number of controls. 
+    """
+
+    case_control_groups = []
+    group_cols = [groupby, replicate_col]
+    if additional_hue is None:
+        hues = [None]
+    else:
+        group_cols.append(additional_hue)
+        hues = df[additional_hue].unique()
+    df = df.groupby(group_cols, observed=True).agg("size").reset_index()
+
+    for hue in hues:
+        if hue is None:
+            tdf = df
+        else:
+            tdf = df.loc[df[additional_hue] == hue, :]
+        cases = tdf.loc[df[groupby] == case_label, :]
+        ncase = cases[0]
+        cases = cases[replicate_col]
+        if control_label is None:
+            controls = tdf.loc[df[groupby] != case_label, :]
+        else:
+            controls = tdf.loc[df[groupby] == control_label, :]
+        ncontrol = controls[0]
+        controls = controls[replicate_col]
+        case_control_groups.append([hue, cases, controls, ncase, ncontrol])
+
+    return case_control_groups
+
+
+def _calculate_imbalance(
+    case_sizes: Union[List, np.ndarray, pd.Series],
```

Comment:
this can be simplified to `typing.Sequence`. 

However, I think you need an array (or Series) for some of the operations below. 
I'm not 100% percent sure, but I think `control_sizes + 0.0001` would fail fi `control_sizes` is a list.